docs/saved_objects/saved_object_repository_factory_design.md

```diff
@@ -0,0 +1,159 @@
+
+The Saved Object Repository is built on top of the OpenSearch client and provides a simplified interface for working with OpenSearch. It uses the Saved Object Serializer to convert saved objects between their internal and external representations. The repository is then being consumed by Saved object client to create scoped saved object client. 
+
+![img](./img/current_saved_object_service_workflow.png)
+
+### Proposed Architecture
+
+- **Approach 1**: In this approach, we just extend the `SavedObjectsRepository` class and override CRUD and other saved object operation in core Dashboards.
+
+    * Pros: 
+        * As we are extending the repository in core saved object service itself, we can reuse the validation and utility functions for other database options. 
+
+
+    * Cons
+        * Changes in core Dashboards : We will be making considerable changes in critical component of Dashboards. 
+        * With this approach, user will have to use the data storage option that we choose.  
+
+- **Approach 2**: Build external plugin and using saved object client wrapper or client factory provider injection mechanism we can build custom object for Postgres or other DB.
+
+    * Pros: 
+        * No changes in core Dashboards. That means we can keep Dashboards as it is with very minimal changes.
+
+
+    * Cons
+        * Code reusability is low. 
+        * Some components of Saved object service such as Serializer, Type registry, interface to create internal and scoped repository are only available during Saved Object Service Start. As per the current architecture, first Saved Object Service Setup → Plugin Setup → Saved Object Service Start → Plugin Start. Some core plugin (e.g. opensearch_dashboards_usage_collection) calls find operation before plugin start and it fails because some components are still not available before plugin start.  
+
+- **Approach 3 (Preferred)**: The proposed architecture will add one more layer of abstraction in Saved Object Service. `The Repository Factory Provider` in Dashboards will be responsible for creating and managing instances of the Repository (e.g. SavedObjectRepository, PostgresRepository, DynamoDBRepository etc.), which is used to interact with the metadata storage that stores the saved objects. Currently we have an repository interface named `ISavedObjectRepository`, and the `SavedObjectRepository` is the implementation, which use an OpenSearch index as its data store. This approach would make the implementation of `ISavedObjectRepository` replaceable by plugin.
```

Comment:
Better add code link to the existing component, using markdown

Reply:
Good idea! I have added code links to existing components.